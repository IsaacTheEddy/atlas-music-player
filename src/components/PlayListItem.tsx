// import data from "../playlist.json";
import { currentSongContext, useCurrentSongContext } from "./context";

const PlayListItem = ({}) => {
  const { playlist, currentSong, setCurrentSong, setLoading, setIsPlaying } =
    useCurrentSongContext();

  const handleClick = (song: any) => {
    if (song.id !== currentSong.id) {
      setLoading?.(true);
      setCurrentSong?.(song);
      setIsPlaying?.(true);
    }
  };

  const songFormatter = (song: number | undefined) => {
    if (!song) {
      return;
    }
    const minutes = Math.floor((song % 3600) / 60);
    const seconds = song % 60;

    const formattedMinutes = String(minutes).padStart(2, "0");
    const formattedSeconds = String(seconds).padStart(2, "0");

    return `${formattedMinutes}:${formattedSeconds}`;
  };

  if (!playlist) {
    return (
      <div className="flex w-100 flex-col">
        <div className="border-double max-md:border-t-8 dark:border-Mocha">
          <h1 className="justify-center border-Lightpink pb-5 font-sans text-2xl font-bold">
            Playlist
          </h1>
        </div>
      </div>
    );
  } else {
    return (
      <div className="flex w-100 flex-col">
        <div className="border-double max-md:border-t-8 dark:border-Mocha">
          <h1 className="justify-center border-Lightpink pb-5 font-sans text-2xl font-bold">
            Playlist
          </h1>
          {playlist.map((song) => {
            if (song.id === currentSong.id) {
              return (
                <>
                  <div
                    className="flex flex-row rounded-md bg-Mocha hover:bg-Mocha"
                    onClick={(e) => handleClick(song)}
                    key={song.id}
                  >
                    <div className="flex w-full flex-col">
                      <h2 className="font-sans font-medium">{song.title}</h2>
                      <p className="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75">
                        {song.artist}
                        {song.cover}
                      </p>
                    </div>
                    <div className="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75">
                      <p>{songFormatter(song.duration)}</p>
                    </div>
                  </div>
                </>
              );
            } else {
              return (
                <>
                  <div
                    className="flex flex-row rounded-md hover:bg-Mocha"
                    onClick={(e) => handleClick(song)}
                    key={song.id}
                  >
                    <div className="flex w-full flex-col">
                      <h2 className="font-sans font-medium">{song.title}</h2>
                      <p className="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75">
                        {song.artist}
                        {song.cover}
                      </p>
                    </div>
                    <div className="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75">
                      <p>{songFormatter(song.duration)}</p>
                    </div>
                  </div>
                </>
              );
            }
          })}
        </div>
      </div>
    );
  }
};

export default PlayListItem;
