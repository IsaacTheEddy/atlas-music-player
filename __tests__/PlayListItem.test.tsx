import { render } from "@testing-library/react";
import { test, expect } from "vitest";
import PlayListItem from "../src/components/PlayListItem";
import {
  currentSongContext,
  useCurrentSongContext,
} from "../src/components/context";

test("Testing for API song", () => {
  const pl = [
    {
      id: "cm3ixp4sy0thg0cmtdzukgg56",
      title: "Painted in Blue",
      artist: "Soul Canvas",
      genre: "Neo-Soul",
      duration: 194,
    },
    {
      id: "soammx6oibpan244my4toqke",
      title: "Tidal Drift",
      artist: "Echoes of the Sea",
      genre: "Ambient",
      duration: 191,
    },
    {
      id: "a6pkp78whsyqdvpb5dxn64ss",
      title: "Fading Shadows",
      artist: "The Emberlight",
      genre: "Alternative Rock",
      duration: 194,
    },
    {
      id: "hdees11mmk6g078ewijlly1r",
      title: "Cosmic Drift",
      artist: "Solar Flare",
      genre: "Psychedelic Rock",
      duration: 240,
    },
    {
      id: "sg9obf34n8qtnpq5t19w1b9a",
      title: "Urban Serenade",
      artist: "Midnight Groove",
      genre: "Jazz/Funk",
      duration: 162,
    },
    {
      id: "u7wkgyiyiz7xuxcueoxzb5d5",
      title: "Whispers in the Wind",
      artist: "Rust & Ruin",
      genre: "Folk",
      duration: 230,
    },
    {
      id: "kp4z671fsjcmuobji96z6p95",
      title: "Electric Fever",
      artist: "Neon Jungle",
      genre: "EDM",
      duration: 141,
    },
    {
      id: "ycvdpo1n76ap0x2u67gy9gab",
      title: "Electric Wildflower",
      artist: "Velvet Ember",
      genre: "Indie Pop",
      duration: 179,
    },
    {
      id: "jxjrthcu47rrpan548evirgu",
      title: "Golden Haze",
      artist: "Velvet Waves",
      genre: "Indie Pop",
      duration: 155,
    },
    {
      id: "kjy3ny0k1jt9p83srhdyg7mb",
      title: "Shatter the Silence",
      artist: "Thunderclap Echo",
      genre: "Rock",
      duration: 224,
    },
  ];
  let { container } = render(
    <currentSongContext.Provider
      value={{
        playlist: pl,
        currentSong: {
          id: "a6pkp78whsyqdvpb5dxn64ss",
          title: "Fading Shadows",
          artist: "The Emberlight",
          genre: "Alternative Rock",
          duration: 194,
          cover:
            "https://utfs.io/f/E9fJnaKtTy1bV09oPkISuh6fWpNsTRlAk1Qj9yqnVzCi32BL",
          song: "https://utfs.io/f/E9fJnaKtTy1ba1N97yFrtlodEaJI0m4wGY9KgyPiUvCf8hMp",
          isPlaying: false,
        },
      }}
    >
      <PlayListItem />
    </currentSongContext.Provider>,
  );

  expect(container).toMatchSnapshot();
});
test("Passing undefined", () => {
  let { container } = render(
    <currentSongContext.Provider
      value={{
        playlist: [{}],
        currentSong: {
          id: "a6pkp78whsyqdvpb5dxn64ss",
          title: "Fading Shadows",
          artist: "The Emberlight",
          genre: "Alternative Rock",
          duration: 194,
          cover:
            "https://utfs.io/f/E9fJnaKtTy1bV09oPkISuh6fWpNsTRlAk1Qj9yqnVzCi32BL",
          song: "https://utfs.io/f/E9fJnaKtTy1ba1N97yFrtlodEaJI0m4wGY9KgyPiUvCf8hMp",
          isPlaying: true,
        },
        loading: true,
      }}
    >
      <PlayListItem />
    </currentSongContext.Provider>,
  );

  expect(container).toMatchSnapshot();
});
