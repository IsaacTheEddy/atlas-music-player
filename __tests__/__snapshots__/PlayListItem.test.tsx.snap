// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Passing undefined 1`] = `
<div>
  <div
    class="flex w-100 flex-col"
  >
    <div
      class="border-double max-md:border-t-8 dark:border-Mocha"
    >
      <h1
        class="justify-center border-Lightpink pb-5 font-sans text-2xl font-bold"
      >
        Playlist
      </h1>
      <div
        class="flex flex-row rounded-md hover:bg-Mocha"
      >
        <div
          class="flex w-full flex-col"
        >
          <h2
            class="font-sans font-medium"
          />
          <p
            class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
          />
        </div>
        <div
          class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
        >
          <p />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing for API song 1`] = `
<div>
  <div
    class="flex w-100 flex-col"
  >
    <div
      class="border-double max-md:border-t-8 dark:border-Mocha"
    >
      <h1
        class="justify-center border-Lightpink pb-5 font-sans text-2xl font-bold"
      >
        Playlist
      </h1>
      <div
        class="flex flex-row rounded-md hover:bg-Mocha"
      >
        <div
          class="flex w-full flex-col"
        >
          <h2
            class="font-sans font-medium"
          >
            Painted in Blue
          </h2>
          <p
            class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
          >
            Soul Canvas
          </p>
        </div>
        <div
          class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
        >
          <p>
            03:14
          </p>
        </div>
      </div>
      <div
        class="flex flex-row rounded-md hover:bg-Mocha"
      >
        <div
          class="flex w-full flex-col"
        >
          <h2
            class="font-sans font-medium"
          >
            Tidal Drift
          </h2>
          <p
            class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
          >
            Echoes of the Sea
          </p>
        </div>
        <div
          class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
        >
          <p>
            03:11
          </p>
        </div>
      </div>
      <div
        class="flex flex-row rounded-md bg-Mocha hover:bg-Mocha"
      >
        <div
          class="flex w-full flex-col"
        >
          <h2
            class="font-sans font-medium"
          >
            Fading Shadows
          </h2>
          <p
            class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
          >
            The Emberlight
          </p>
        </div>
        <div
          class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
        >
          <p>
            03:14
          </p>
        </div>
      </div>
      <div
        class="flex flex-row rounded-md hover:bg-Mocha"
      >
        <div
          class="flex w-full flex-col"
        >
          <h2
            class="font-sans font-medium"
          >
            Cosmic Drift
          </h2>
          <p
            class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
          >
            Solar Flare
          </p>
        </div>
        <div
          class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
        >
          <p>
            04:00
          </p>
        </div>
      </div>
      <div
        class="flex flex-row rounded-md hover:bg-Mocha"
      >
        <div
          class="flex w-full flex-col"
        >
          <h2
            class="font-sans font-medium"
          >
            Urban Serenade
          </h2>
          <p
            class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
          >
            Midnight Groove
          </p>
        </div>
        <div
          class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
        >
          <p>
            02:42
          </p>
        </div>
      </div>
      <div
        class="flex flex-row rounded-md hover:bg-Mocha"
      >
        <div
          class="flex w-full flex-col"
        >
          <h2
            class="font-sans font-medium"
          >
            Whispers in the Wind
          </h2>
          <p
            class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
          >
            Rust & Ruin
          </p>
        </div>
        <div
          class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
        >
          <p>
            03:50
          </p>
        </div>
      </div>
      <div
        class="flex flex-row rounded-md hover:bg-Mocha"
      >
        <div
          class="flex w-full flex-col"
        >
          <h2
            class="font-sans font-medium"
          >
            Electric Fever
          </h2>
          <p
            class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
          >
            Neon Jungle
          </p>
        </div>
        <div
          class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
        >
          <p>
            02:21
          </p>
        </div>
      </div>
      <div
        class="flex flex-row rounded-md hover:bg-Mocha"
      >
        <div
          class="flex w-full flex-col"
        >
          <h2
            class="font-sans font-medium"
          >
            Electric Wildflower
          </h2>
          <p
            class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
          >
            Velvet Ember
          </p>
        </div>
        <div
          class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
        >
          <p>
            02:59
          </p>
        </div>
      </div>
      <div
        class="flex flex-row rounded-md hover:bg-Mocha"
      >
        <div
          class="flex w-full flex-col"
        >
          <h2
            class="font-sans font-medium"
          >
            Golden Haze
          </h2>
          <p
            class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
          >
            Velvet Waves
          </p>
        </div>
        <div
          class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
        >
          <p>
            02:35
          </p>
        </div>
      </div>
      <div
        class="flex flex-row rounded-md hover:bg-Mocha"
      >
        <div
          class="flex w-full flex-col"
        >
          <h2
            class="font-sans font-medium"
          >
            Shatter the Silence
          </h2>
          <p
            class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
          >
            Thunderclap Echo
          </p>
        </div>
        <div
          class="font-sans font-medium text-Mocha/75 dark:text-Lightsoftie/75"
        >
          <p>
            03:44
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
